src/System.IO.Pipelines.Networking.Tls/OpenSslConnectionContext.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System.IO.Pipelines.Networking.Tls.Internal.OpenSsl;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines.Networking.Tls
+{
+    public class OpenSslConnectionContext : ISecureContext
+    {
+        private static readonly Task CachedTask = Task.FromResult(0);
+
+        private readonly OpenSslSecurityContext _securityContext;
+        private int _headerSize = 5; //5 is the minimum (1 for frame type, 2 for version, 2 for frame size)
+        private int _trailerSize = 16; //Used for MAC storage
+        private bool _readyToSend;
+        private IntPtr _ssl;
+        private readonly InteropBio.BioHandle _readBio;
+        private readonly InteropBio.BioHandle _writeBio;
+        private ApplicationProtocols.ProtocolIds _negotiatedProtocol;
+
+        public OpenSslConnectionContext(OpenSslSecurityContext securityContext, IntPtr ssl)
+        {
+            _ssl = ssl;
+            _securityContext = securityContext;
+            _writeBio = InteropBio.BIO_new(CustomBio.Custom());
+            _readBio = InteropBio.BIO_new(CustomBio.Custom());
+
+            Interop.SSL_set_bio(_ssl, _readBio, _writeBio);
+            if (IsServer)
+            {
+                Interop.SSL_set_accept_state(_ssl);
+            }
+            else
+            {
+                Interop.SSL_set_connect_state(_ssl);
+            }
+        }
+
+        public bool IsServer => _securityContext.IsServer;
+
+        public int HeaderSize
+        {
+            get { return _headerSize; }
+            set { _headerSize = value; }
+        }
+
+        public int TrailerSize
+        {
+            get { return _trailerSize; }
+            set { _trailerSize = value; }
+        }
+
+        public ApplicationProtocols.ProtocolIds NegotiatedProtocol => _negotiatedProtocol;
+        public bool ReadyToSend => _readyToSend;
+        public CipherInfo CipherInfo => _ssl != IntPtr.Zero ? Interop.GetCipherInfo(_ssl) : default(CipherInfo);
+
+        public unsafe Task DecryptAsync(ReadableBuffer encryptedData, IPipelineWriter decryptedPipeline)
+        {
+            CustomBio.SetReadBufferPointer(_readBio, ref encryptedData);
+
+            var decryptedData = decryptedPipeline.Alloc();
+            var result = 1;
+            while (result > 0)
+            {
+                void* memPtr;
+                decryptedData.Ensure(1024);
+
+                decryptedData.Memory.TryGetPointer(out memPtr);
+                result = Interop.SSL_read(_ssl, memPtr, decryptedData.Memory.Length);
+                if (result > 0)
```

Comment:
On the decryptedData.Memory.TryGetPointer? Not actually sure how this ever fails? @davidfowl might be able to answer that.